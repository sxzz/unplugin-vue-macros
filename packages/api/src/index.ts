export { MagicString } from '@vue-macros/common'

// WIP
// export * from './vue'
export * from './ts'
export * from './utils'
