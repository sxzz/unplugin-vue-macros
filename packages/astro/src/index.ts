import { type AstroIntegration } from 'astro'
import { type Options } from 'unplugin-vue-macros'
import VitePlugin from 'unplugin-vue-macros/vite'

const name = 'unplugin-vue-macros'

export default function VueMacrosAstroIntegration(
  userOptions: Options = {}
): AstroIntegration {
  return {
    name,
    hooks: {
      'astro:config:setup': ({ updateConfig }) => {
        updateConfig({
          vitePlugin: [VitePlugin(userOptions)],
        })
      },
    },
  }
}
