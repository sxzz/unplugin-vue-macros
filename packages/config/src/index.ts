export * from './config-worker'
export * from './config'
export * from './define'
export * from './options'
