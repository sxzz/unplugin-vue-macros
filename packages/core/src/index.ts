import { compileScript } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import { DEFINE_EMITS, DEFINE_MODEL, DEFINE_PROPS } from './constants'
import { isCallOf, parseSFC } from './utils'
import type {
  LVal,
  Node,
  Statement,
  TSInterfaceBody,
  TSTypeLiteral,
} from '@babel/types'

export const transform = (code: string, filename: string) => {
  let hasDefineProps = false
  let hasDefineEmits = false
  let hasDefineModelCall = false
  let propsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let emitsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let modelTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let modelIdentifier: string | undefined

  function processDefinePropsOrEmits(node: Node) {
    let type: 'props' | 'emits'
    if (isCallOf(node, DEFINE_PROPS)) {
      type = 'props'
    } else if (isCallOf(node, DEFINE_EMITS)) {
      type = 'emits'
    } else {
      return false
    }
    const fnName = type === 'props' ? DEFINE_PROPS : DEFINE_EMITS

    if (type === 'props') hasDefineProps = true
    else hasDefineEmits = true

    if (node.arguments[0]) {
      throw new SyntaxError('error: 2')
    }

    const typeDeclRaw = node.typeParameters?.params?.[0]
    if (!typeDeclRaw) {
      throw new SyntaxError('Error: 3')
    }
    const typeDecl = resolveQualifiedType(
      typeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!typeDecl) {
      throw new Error(
        `type argument passed to ${fnName}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    if (type === 'props') propsTypeDecl = typeDecl
    else emitsTypeDecl = typeDecl
  }

  function processDefineModel(node: Node, declId: LVal) {
    if (!isCallOf(node, DEFINE_MODEL)) {
      return false
    }

    if (hasDefineModelCall) {
      throw new SyntaxError(`duplicate ${DEFINE_MODEL}() call`)
    }
    hasDefineModelCall = true

    const propsTypeDeclRaw = node.typeParameters?.params[0]
    if (!propsTypeDeclRaw) {
      throw new SyntaxError('Error: 1')
    }
    modelTypeDecl = resolveQualifiedType(
      propsTypeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!modelTypeDecl) {
      throw new Error(
        `type argument passed to ${DEFINE_MODEL}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    modelIdentifier = scriptSetup.loc.source.slice(declId.start!, declId.end!)
  }

  function resolveQualifiedType(
    node: Node,
    qualifier: (node: Node) => boolean
  ) {
    if (qualifier(node)) {
      return node
    }
    if (
      node.type === 'TSTypeReference' &&
      node.typeName.type === 'Identifier'
    ) {
      const refName = node.typeName.name
      const isQualifiedType = (node: Node): Node | undefined => {
        if (
          node.type === 'TSInterfaceDeclaration' &&
          node.id.name === refName
        ) {
          return node.body
        } else if (
          node.type === 'TSTypeAliasDeclaration' &&
          node.id.name === refName &&
          qualifier(node.typeAnnotation)
        ) {
          return node.typeAnnotation
        } else if (node.type === 'ExportNamedDeclaration' && node.declaration) {
          return isQualifiedType(node.declaration)
        }
      }
      const body = sfc.scriptSetup!.scriptSetupAst!
      for (const node of body) {
        const qualified = isQualifiedType(node)
        if (qualified) {
          return qualified
        }
      }
    }
  }

  function extractRuntimeProps(
    node: TSTypeLiteral | TSInterfaceBody
  ): Record<string, string> {
    const members = node.type === 'TSTypeLiteral' ? node.members : node.body
    const map: Record<string, string> = {}
    for (const m of members) {
      if (
        (m.type === 'TSPropertySignature' || m.type === 'TSMethodSignature') &&
        m.key.type === 'Identifier'
      ) {
        const value = scriptSetup.loc.source.slice(
          m.typeAnnotation!.start!,
          m.typeAnnotation!.end!
        )
        map[m.key.name] = value
      }
    }
    return map
  }

  if (!code.includes(DEFINE_MODEL)) return

  const sfc = parseSFC(code, filename)
  if (!sfc.scriptSetup) return

  if (!sfc.scriptSetup.scriptSetupAst) {
    sfc.scriptSetup = compileScript(sfc, {
      id: filename,
    })
  }
  const { scriptSetup } = sfc
  const startOffset = scriptSetup.loc.start.offset
  // const endOffset = scriptSetup.loc.end.offset

  const s = new MagicString(code)

  for (const node of scriptSetup.scriptSetupAst as Statement[]) {
    if (node.type === 'ExpressionStatement') {
      processDefinePropsOrEmits(node.expression)
    }

    if (node.type === 'VariableDeclaration' && !node.declare) {
      const total = node.declarations.length
      let left = total

      for (let i = 0; i < total; i++) {
        const decl = node.declarations[i]
        if (decl.init) {
          processDefinePropsOrEmits(decl.init)
          processDefineModel(decl.init, decl.id)

          if (hasDefineModelCall) {
            if (left === 1) {
              s.remove(node.start! + startOffset, node.end! + startOffset)
            } else {
              let start = decl.start! + startOffset
              let end = decl.end! + startOffset
              if (i < total - 1) {
                // not the last one, locate the start of the next
                end = node.declarations[i + 1].start! + startOffset
              } else {
                // last one, locate the end of the prev
                start = node.declarations[i - 1].end! + startOffset
              }
              s.remove(start, end)
              left--
            }
          }
        }
      }
    }
  }

  if (!modelTypeDecl) return

  if (modelTypeDecl.type !== 'TSTypeLiteral') {
    throw new SyntaxError(
      `type argument passed to ${DEFINE_MODEL}() must be a literal type, or a reference to an interface or literal type.`
    )
  }

  const map = extractRuntimeProps(modelTypeDecl)

  const propsText = Object.entries(map)
    .map(([key, value]) => `${key}${value}`)
    .join('\n')

  const emitsText = Object.entries(map)
    .map(([key, value]) => `(evt: 'update:${key}', value${value}): void`)
    .join('\n')

  if (hasDefineProps) {
    s.appendLeft(startOffset + propsTypeDecl!.start! + 1, `${propsText}\n`)
  } else {
    s.appendLeft(
      startOffset,
      `const ${modelIdentifier} = defineProps<{
  ${propsText}
}>();\n`
    )
  }
  if (hasDefineEmits) {
    s.appendLeft(startOffset + emitsTypeDecl!.start! + 1, `${emitsText}\n`)
  } else {
    s.appendLeft(
      startOffset,
      `const _${DEFINE_MODEL}_emit = defineEmits<{
  ${emitsText}
}>();\n`
    )
  }

  return {
    code: s.toString(),
    get map() {
      return s.generateMap()
    },
  }
}
