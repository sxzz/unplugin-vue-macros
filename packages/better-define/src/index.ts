import process from 'node:process'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBetterDefine } from './core'
import type { PluginContext } from 'rollup'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude: filter,
      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (error: unknown) {
          this.warn(`${name} ${error}`)
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
export default plugin
