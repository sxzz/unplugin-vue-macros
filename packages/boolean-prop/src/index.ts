import { getVuePluginApi, type VuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBooleanProp, type Options } from './core/index'
import type { Plugin } from 'vite'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options),
      )
    },
  }
}

const plugin: {
  rollup: typeof rollup
  vite: typeof rollup
} = {
  rollup,
  vite: rollup,
}
export default plugin
