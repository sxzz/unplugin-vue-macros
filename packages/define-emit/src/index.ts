import process from 'node:process'
import { type UnpluginContextMeta, createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  FilterFileType,
  type MarkRequired,
  createFilter,
  detectVueVersion,
  getFilterPattern,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { transformDefineEmit } from './core'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)
    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude: filter,
      transform: transformDefineEmit,

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
