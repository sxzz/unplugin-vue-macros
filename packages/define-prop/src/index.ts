import process from 'node:process'
import { type UnpluginContextMeta, createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  FilterFileType,
  type MarkRequired,
  createFilter,
  detectVueVersion,
  getFilterPattern,
  normalizePath,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Edition, transformDefineProp } from './core'
import { helperCode, helperId } from './core/helper'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
  edition?: Edition
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction' | 'edition'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    edition: 'kevinEdition',
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)
    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineProp(
          code,
          id,
          options.edition,
          options.isProduction,
        )
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
