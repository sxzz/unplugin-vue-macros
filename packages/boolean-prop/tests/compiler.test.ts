import { compileTemplate } from '@vue/compiler-sfc'
import { describe, expect, test } from 'vitest'
import { transformBooleanProp } from '../src'

function compile(code: string, negativePrefix?: string) {
  return compileTemplate({
    source: code,
    filename: 'anonymous.vue',
    id: 'xxx',
    compilerOptions: {
      nodeTransforms: [
        transformBooleanProp({
          negativePrefix,
        }),
      ],
    },
  })
}

describe('compiler', () => {
  test('basic', () => {
    const original = compile(`<input :checked="true" />`)
    const sugar = compile(`<input checked />`)
    expect(sugar.code).toBe(original.code)
    expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
  })

  test('false prop', () => {
    const original = compile(`<input :checked="false" />`)
    const sugar = compile(`<input !checked />`)
    expect(sugar.code).toBe(original.code)
    expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
  })

  test('false prop prefix', () => {
    const original = compile(`<input :checked="false" !w-5 />`, '~')
    const sugar = compile(`<input ~checked !w-5 />`, '~')
    expect(sugar.code).toBe(original.code)
    expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
  })
})
