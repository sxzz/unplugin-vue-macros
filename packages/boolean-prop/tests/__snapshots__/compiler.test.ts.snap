// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler > basic 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "checked",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "source": "checked",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 3,
    "content": "true",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "source": "true",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "source": "checked",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "bind",
  "type": 7,
}
`;

exports[`compiler > false prop 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "checked",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "source": "!checked",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 3,
    "content": "false",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "source": "false",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "source": "!checked",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 8,
    },
  },
  "modifiers": [],
  "name": "bind",
  "type": 7,
}
`;

exports[`compiler > false prop prefix 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "checked",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "source": "~checked",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 3,
    "content": "false",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "source": "false",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "source": "~checked",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 8,
    },
  },
  "modifiers": [],
  "name": "bind",
  "type": 7,
}
`;
